File: SpeedTestGameClient/core/src/main/java/si/um/feri/speedii/dao/http/HttpMeasurement.java
package si.um.feri.speedii.dao.http;

import io.github.cdimascio.dotenv.Dotenv;
import okhttp3.*;
import si.um.feri.speedii.classes.Location;
import si.um.feri.speedii.classes.Measurement;
import si.um.feri.speedii.classes.SessionManager;
import si.um.feri.speedii.classes.Type;
import si.um.feri.speedii.classes.User;
import si.um.feri.speedii.dao.MeasurementCRUD;

import org.bson.types.ObjectId;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class HttpMeasurement implements MeasurementCRUD {

    private final OkHttpClient client;
    private final String ip;
    private final SessionManager sessionManager;

    public HttpMeasurement(SessionManager sessionManager) {
        this.sessionManager = sessionManager;
        this.client = new OkHttpClient();

        Dotenv dotenv = Dotenv.configure().filename(".env").load();
        this.ip = dotenv.get("IP");
    }


    //TESTED
    @Override
    public List<Measurement> getByUser(User user) throws IOException {
        String url = ip + "/measurements/user/" + user.getId();
        Request request = new Request.Builder()
            .url(url)
            .get()
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Failed to execute request: " + response.code());
                return new ArrayList<>();
            }

            String responseBody = response.body().string();
            JSONArray jsonArray = new JSONArray(responseBody);
            List<Measurement> measurements = new ArrayList<>();

            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                measurements.add(parseMeasurement(jsonObject));
            }

            return measurements;
        }
    }

    //TESTED
    @Override
    public List<Measurement> getByTimeFrame(LocalDateTime start, LocalDateTime end) throws IOException {
        String startTime = start.format(DateTimeFormatter.ISO_DATE_TIME);
        String endTime = end.format(DateTimeFormatter.ISO_DATE_TIME);
        String url = ip + "/measurements/timeframe/" + startTime + "/" + endTime;
        Request request = new Request.Builder()
            .url(url)
            .get()
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Failed to execute request: " + response.code());
                return new ArrayList<>();
            }

            String responseBody = response.body().string();
            JSONArray jsonArray = new JSONArray(responseBody);
            List<Measurement> measurements = new ArrayList<>();

            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                measurements.add(parseMeasurement(jsonObject));
            }

            return measurements;
        }
    }

    //TESTED
    @Override
    public Measurement getById(ObjectId id) throws IOException {
        String url = ip + "/measurements/" + id;
        Request request = new Request.Builder()
            .url(url)
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Failed to execute request: " + response.code());
                return null;
            }

            String responseBody = response.body().string();
            JSONObject jsonObject = new JSONObject(responseBody);
            return parseMeasurement(jsonObject);
        }
    }

    //TESTED
    @Override
    public List<Measurement> getAll() throws IOException {
        String url = ip + "/measurements";
        Request request = new Request.Builder()
            .url(url)
            .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Failed to execute request: " + response.code());
                return new ArrayList<>();
            }

            String responseBody = response.body().string();
            JSONArray jsonArray = new JSONArray(responseBody);
            List<Measurement> measurements = new ArrayList<>();

            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                measurements.add(parseMeasurement(jsonObject));
            }

            return measurements;
        }
    }

    //TESTED
    @Override
    public boolean insert(Measurement obj) throws IOException {
        JSONObject requestBody = new JSONObject();
        requestBody.put("speed", obj.getSpeed());
        requestBody.put("type", obj.getType().name());
        requestBody.put("provider", obj.getProvider());
        requestBody.put("time", obj.getTime().format(DateTimeFormatter.ISO_DATE_TIME));

        JSONObject locationObject = new JSONObject();
        locationObject.put("type", "Point");
        JSONArray coordinatesArray = new JSONArray();
        for (double coordinate : obj.getLocation().coordinates) {
            coordinatesArray.put(coordinate);
        }
        locationObject.put("coordinates", coordinatesArray);

        requestBody.put("location", locationObject);
        if (obj.getUser() != null) {
            requestBody.put("measuredBy", obj.getUser().getId());
        }

        Request request = new Request.Builder()
            .url(ip + "/measurements")
            .post(RequestBody.create(requestBody.toString(), MediaType.get("application/json")))
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        }
    }

    @Override
    public boolean insertMany(List<Measurement> list) {
        JSONArray jsonArray = new JSONArray();
        for (Measurement measurement : list) {
            JSONObject jsonObject = new JSONObject();
            jsonObject.put("speed", measurement.getSpeed());
            jsonObject.put("type", measurement.getType().name());
            jsonObject.put("provider", measurement.getProvider());
            jsonObject.put("time", measurement.getTime().format(DateTimeFormatter.ISO_DATE_TIME));
            JSONObject locationObject = new JSONObject();
            locationObject.put("type", "Point");
            locationObject.put("coordinates", new JSONArray(measurement.getLocation().coordinates));
            jsonObject.put("location", locationObject);
            if (measurement.getUser() != null) {
                jsonObject.put("measuredBy", measurement.getUser().getId());
            }
            jsonArray.put(jsonObject);
        }

        JSONObject jsonMeasurements = new JSONObject();
        jsonMeasurements.put("measurements", jsonArray);

        RequestBody requestBody = RequestBody.create(
            jsonMeasurements.toString(), MediaType.get("application/json"));

        Request request = new Request.Builder()
            .url(ip + "/measurements/createMany")
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .post(requestBody)  // Here we use `post` with `RequestBody`
            .build();

        try {
            Response response = client.newCall(request).execute();
            return response.isSuccessful();  // Returns true if the request was successful
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    @Override
    public boolean update(Measurement obj) throws IOException {
        String url = ip + "/measurements/" + obj.getId();
        JSONObject requestBody = new JSONObject();
        requestBody.put("speed", obj.getSpeed());
        requestBody.put("type", obj.getType().name());
        requestBody.put("provider", obj.getProvider());
        requestBody.put("time", obj.getTime().format(DateTimeFormatter.ISO_DATE_TIME));

        JSONObject locationObject = new JSONObject();
        locationObject.put("type", "Point");
        locationObject.put("coordinates", obj.getLocation().coordinates);

        requestBody.put("location", locationObject);
        if (obj.getUser() != null) {
            requestBody.put("measuredBy", obj.getUser().getId());
        }

        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .put(RequestBody.create(requestBody.toString(), MediaType.get("application/json")))
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        }
    }


    //TESTED
    @Override
    public boolean delete(Measurement obj) throws IOException {
        String url = ip + "/measurements/" + obj.getId();
        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .delete()
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        }
    }

    private Measurement parseMeasurement(JSONObject jsonObject) {
        long speed = jsonObject.getLong("speed");
        Type type = Type.valueOf(jsonObject.getString("type"));
        String provider = jsonObject.getString("provider");

        String timeString = jsonObject.getString("time");
        DateTimeFormatter formatter = DateTimeFormatter.ISO_DATE_TIME;
        LocalDateTime time = LocalDateTime.parse(timeString, formatter);

        JSONObject locationObject = jsonObject.getJSONObject("location");
        List<Double> coordinates = new ArrayList<>();
        JSONArray coordinatesArray = locationObject.getJSONArray("coordinates");
        for (int i = 0; i < coordinatesArray.length(); i++) {
            coordinates.add(coordinatesArray.getDouble(i));
        }
        Location location = new Location(locationObject.getString("type"), coordinates);

        User user = null;
        if (!jsonObject.isNull("measuredBy")) {
            user = parseUser(jsonObject.getJSONObject("measuredBy"));
        }

        ObjectId id = !jsonObject.isNull("_id") && jsonObject.get("_id") instanceof String ?
            new ObjectId(jsonObject.getString("_id")) : new ObjectId();

        return new Measurement(speed, type, provider, location, time, user, id);
    }

    private User parseUser(JSONObject jsonObject) {
        String username = jsonObject.getString("username");
        String password = jsonObject.getString("password");
        String email = jsonObject.getString("email");
        boolean admin = jsonObject.getBoolean("admin");
        ObjectId id = new ObjectId(jsonObject.getString("_id"));
        return new User(username, password, email, admin, id);
    }
}

File: MobileApp/app/speedtest/src/main/java/sosteric/pora/speedtest/dao/http/HttpMeasurement.kt
package dao.http

import Location
import Measurment
import SessionManager
import User
import dao.MeasurementCrud
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.bson.types.ObjectId
import org.json.JSONArray
import org.json.JSONObject
import sosteric.pora.speedtest.Type
import java.io.File
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

class HttpMeasurement(val sessionManager: SessionManager): MeasurementCrud {
    val client = OkHttpClient()
    var ip = "http://20.160.43.196:3000"

    override fun getByUser(user: User): List<Measurment> {
        val url = "${ip}/measurements/user/${user.id}"
        val request = Request.Builder()
            .url(url)
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                println("Failed to execute request: ${response.code}")
                return emptyList()
            }

            val responseBody = response.body?.string()
            responseBody?.let {
                val jsonArray = JSONArray(it)
                val measurements = mutableListOf<Measurment>()

                for (i in 0 until jsonArray.length()) {
                    val jsonObject = jsonArray.getJSONObject(i)

                    measurements.add(parseMeasurement(jsonObject))
                }

                return measurements
            }

            return emptyList()
        }

    }

    override fun getByTimeFrame(start: LocalDateTime, end: LocalDateTime): List<Measurment> {
        val startTime = start.format(DateTimeFormatter.ISO_DATE_TIME)
        val endTime = end.format(DateTimeFormatter.ISO_DATE_TIME)
        val url = "${ip}/measurements/timeframe/${startTime}/${endTime}"
        val request = Request.Builder()
            .url(url)
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                println("Failed to execute request: ${response.code}")
                return emptyList()
            }

            val responseBody = response.body?.string()
            responseBody?.let {
                val jsonArray = JSONArray(it)
                val measurements = mutableListOf<Measurment>()

                for (i in 0 until jsonArray.length()) {
                    val jsonObject = jsonArray.getJSONObject(i)

                    measurements.add(parseMeasurement(jsonObject))
                }

                return measurements
            }

            return emptyList()
        }
    }


    override fun getById(id: ObjectId): Measurment? {
        val url = "${ip}/measurements/${id}"
        val request = Request.Builder()
            .url(url)
            .build()

        client.newCall(request).execute().use { response ->
            // Check if the request was successful (HTTP 200 OK)
            if (!response.isSuccessful) {
                // If not, handle the error
                println("Failed to execute request: ${response.code}")
                return null
            }

            // Get the response body as a string
            val responseBody = response.body?.string()

            responseBody?.let {
                val jsonObject = JSONObject(it)
                return parseMeasurement(jsonObject)
            }

            return null
        }

    }

    override fun getAll(): List<Measurment>  {
        val url = "${ip}/measurements"
        val request = Request.Builder()
            .url(url)
            .build()

        client.newCall(request).execute().use { response ->
            // Check if the request was successful (HTTP 200 OK)
            if (!response.isSuccessful) {
                // If not, handle the error
                println("Failed to execute request: ${response.code}")
                return emptyList()
            }

            // Get the response body as a string
            val responseBody = response.body?.string()

            responseBody?.let {
                val jsonArray = JSONArray(it)
                val measurements = mutableListOf<Measurment>()

                for (i in 0 until jsonArray.length()) {
                    val jsonObject = jsonArray.getJSONObject(i)

                    measurements.add(parseMeasurement(jsonObject))
                }

                return measurements
            }

            return emptyList()
        }

    }
    override fun insert(obj: Measurment): Boolean {
        println(obj.user)
        val requestBody = JSONObject().apply {
            put("speed", obj.speed)
            put("type", obj.type.name)
            put("provider", obj.provider)
            put("time", obj.time.format(DateTimeFormatter.ISO_DATE_TIME))
            put("location", JSONObject().apply {
                put("type", "Point")
                put("coordinates", obj.location.coordinates)
            })
            obj.user?.let { put("measuredBy", it.id) }
        }

        val request = Request.Builder()
            .url("${ip}/measurements")
            .post(requestBody.toString().toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(request).execute().use { response ->
            return response.isSuccessful
        }
    }

    override fun insertMany(list: List<Measurment>): Boolean {
        val jsonArray = JSONArray()
        list.forEach { measurement ->
            val jsonObject = JSONObject().apply {
                put("speed", measurement.speed)
                put("type", measurement.type.name)
                put("provider", measurement.provider)
                put("time", measurement.time.format(DateTimeFormatter.ISO_DATE_TIME))
                put("location", JSONObject().apply {
                    put("type", "Point")
                    put("coordinates", measurement.location.coordinates)
                })
                measurement.user?.let { put("measuredBy", it.id) }
            }
            jsonArray.put(jsonObject)
        }

        val jsonMeasurements = JSONObject().apply {
            put("measurements", jsonArray)
        }
        val requestBody = jsonMeasurements.toString().toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url("${ip}/measurements/createMany")
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .post(requestBody)
            .build()

        client.newCall(request).execute().use { response ->
            return response.isSuccessful
        }
    }

    override fun update(obj: Measurment): Boolean {
        val url = "${ip}/measurements/${obj.id}"
        val requestBody = JSONObject().apply {
            put("speed", obj.speed)
            put("type", obj.type.name)
            put("provider", obj.provider)
            put("time", obj.time.format(DateTimeFormatter.ISO_DATE_TIME))
            put("location", JSONObject().apply {
                put("type", "Point")
                put("coordinates", obj.location.coordinates)
            })
            obj.user?.let { put("measuredBy", it.id) }
        }
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .put(requestBody.toString().toRequestBody("application/json".toMediaType()))
            .build()

        try {
            val response = client.newCall(request).execute()
            return response.isSuccessful
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }
    }

    override fun delete(obj: Measurment): Boolean {
        val url = "${ip}/measurements/${obj.id}"
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .delete()
            .build()

        try {
            val response = client.newCall(request).execute()
            return response.isSuccessful
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }
    }

    private fun parseMeasurement(jsonObject: JSONObject) : Measurment{
        val speed = jsonObject.getLong("speed")
        val type = Type.valueOf(jsonObject.getString("type"))
        val provider = jsonObject.getString("provider")

        val timeString = jsonObject.getString("time")
        val formatter = DateTimeFormatter.ISO_DATE_TIME
        val time = LocalDateTime.parse(timeString, formatter)

        val locationObject = jsonObject.getJSONObject("location")
        val location = Location(
            type = locationObject.getString("type"),
            coordinates = locationObject.getJSONArray("coordinates")
                .let { coordinatesArray ->
                    List<Double>(coordinatesArray.length()) { coordinatesArray.getDouble(it) }
                }
        )
        val user = if (!jsonObject.isNull("measuredBy")) {
            parseUser(jsonObject.getJSONObject("measuredBy"))
        } else {
            null
        }
        val id = if (!jsonObject.isNull("_id") && jsonObject.get("_id") is String) {
            ObjectId(jsonObject.getString("_id"))
        } else {
            ObjectId()
        }

        return Measurment(speed, type, provider, location, time, user, id)
    }
    fun parseUser(jsonObject: JSONObject) : User {
        val username = jsonObject.getString("username")
        val password = jsonObject.getString("password")
        val email = jsonObject.getString("email")
        val admin = jsonObject.getBoolean("admin")
        val id = ObjectId(jsonObject.getString("_id"))
        return User(username, password, email, admin, id)
    }

}